modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java

```diff
@@ -625,6 +633,42 @@ public boolean visitFunction(Function function, Object extraData) throws IOExcep
         return true;
     }
 
+    private void encodeJsonPointer(Function jsonPointer, Object extraData) throws IOException {
+        Expression json = getParameter(jsonPointer, 0, true);
+        Expression pointer = getParameter(jsonPointer, 1, true);
+        if (json instanceof PropertyName && pointer instanceof Literal) {
+            // if not a string need to cast the json attribute
+            boolean needCast =
+                    extraData != null
+                            && extraData instanceof Class
+                            && !extraData.equals(String.class);
+
+            if (needCast) out.write('(');
+            json.accept(delegate, null);
+            out.write(" ::json ");
+            String strPointer = ((Literal) pointer).getValue().toString();
```

Comment:
In general you cannot assume it's going to be a Literal, functions must work with whatever expression they are given as input. The existing function, without the postgresql encoding, would work even if the pointer was specified with a complex expression, possibly attribute based.

If you check [SQLDialect](https://github.com/geotools/geotools/blob/7529923af1143cfe25cd475bb4f05c11fca83957/modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java#L1259), it gets called for splitting filters. For this case, I suggest to override it, and check if there is a pointer function, and if the function is called in a "static" way. Static does not mean "literal" directly, it just means that the expression tree referenced by  "pointer" :

- does not use any attribute
- does not use any VolatileFunction

If those conditions are met, you can include the function in the capabilities, otherwise, you exclude it from them and have the splitter move it to the post-filter.
I think (please verify) that the JDBCDataStore code makes the filter go through a SimplfyingFilterVisitor, which is already doing this job for you:
https://github.com/geotools/geotools/blob/029dc5c8ccc503de3cfd7f4d1c939d8056c260c4/modules/library/main/src/main/java/org/geotools/filter/visitor/SimplifyingFilterVisitor.java#L453

If so, all you need to do is to write a DefaultFilterVisitor, if it finds a JSONPointerFunction, it means it was not eliminated by the simplifier, and thus it's dynamic, cannot be encoded in SQL.

Reply:
thanks for the point and the explanation, fixed